sdk/core/Azure.Core/src/AzureKeyCredential.cs

```diff
@@ -54,7 +61,18 @@ public string Key
         public void Update(string key)
         {
             Argument.AssertNotNullOrEmpty(key, nameof(key));
+
+            var oldKey = Key;
             Key = key;
+
+            var changed = KeyChanged;
+            if (changed != null)
+            {
+                if (!string.Equals(oldKey, key, StringComparison.Ordinal))
+                {
+                    changed(this, Key);
+                }
+            }
```

Comment:
I feel like there is a bit of a race condition here, because if I subscribe to this event the order the event is raised is not necessarily the order that the key is updated in. This isn't necessarily a problem, but it might be counter to users expectations.  I see a couple of ways you could possibly deal with this.

1. Don't pass the updated value of the key to the event. The subscriber must then go back to the key to get the most up to date value (even if multiple callers had updated, the subscriber would still get the latest value)
2. Lock around the key update and raising this event. (Then we're guaranteed nothing else updated before a subscriber handles the event, but subscribers could essentially block other updates unfortunately)

Reply:
I will lock. 
Having said that, updating keys on different threads concurrently will be very hard to make reliable. I suspect now the user will have a race. They will transform the string to a byte array and might do it in the wrong order.

Reply:
Actually, to lock I would have to add a sync root to AKC. Do we worry about it?

Reply:
Damn it, Scott.   I saw this comment just as I was about to sign off yesterday and spend the evening feeling bad about myself for not thinking of that scenario...